t/protocol_directnic.pl

```diff
@@ -0,0 +1,197 @@
+    },
+    {
+        desc => 'IPv4, single host, unexpected response',
+        cfg => {h1 => {urlv4 => "$hostname/unexpected/path/", wantipv4 => '192.0.2.1'}},
+        wantstatus => {h1 => {}},
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/400 Bad Request/},
+        ],
+    },
+    {
+        desc => 'IPv4, single host, no urlv4',
+        cfg => {h1 => {wantipv4 => '192.0.2.1'}},
+        wantstatus => {h1 => {}},
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/missing urlv4 option/},
+        ],
+    },
+    {
+        desc => 'IPv6, single host, good',
+        cfg => {h1 => {urlv6 => "$hostname/dns/gateway/abc/", wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv6' => 'good', 'ipv6' => '2001:db8::1', 'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, good',
+        cfg => {h1 => {
+            urlv4 => "$hostname/dns/gateway/abc/",
+            urlv6 => "$hostname/dns/gateway/def/",
+            wantipv4 => '192.0.2.1',
+            wantipv6 => '2001:db8::1',
+        }},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'good', 'ipv4' => '192.0.2.1',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv4/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+    {
+        desc => 'IPv4 and IPv6, single host, mixed success',
+        cfg => {h1 => {
+            urlv4 => "$hostname/dns/gateway/bad_token/",
+            urlv6 => "$hostname/dns/gateway/def/",
+            wantipv4 => '192.0.2.1',
+            wantipv6 => '2001:db8::1',
+        }},
+        wantips => {h1 => {wantipv4 => '192.0.2.1', wantipv6 => '2001:db8::1'}},
+        wantstatus => {
+            h1 => {'status-ipv4' => 'failed',
+                'status-ipv6' => 'good', 'ipv6' => '2001:db8::1',
+                'mtime' => $ddclient::now},
+        },
+        wantlogs => [
+            {label => 'FAILED', ctx => ['h1'], msg => qr/There was an error updating your record/},
+            {label => 'SUCCESS', ctx => ['h1'], msg => qr/IPv6/},
+        ],
+    },
+);
+
+for my $tc (@test_cases) {
+    diag('==============================================================================');
+    diag("Starting test: $tc->{desc}");
+    diag('==============================================================================');
+    local $ddclient::globals{debug} = 1;
+    local $ddclient::globals{verbose} = 1;
+    my $l = Logger->new($ddclient::_l);
+    local %ddclient::config;
+    $ddclient::config{$_} = {
+        protocol => 'directnic',
+        host => $_,
+        %{$tc->{cfg}{$_}},
+    } for keys(%{$tc->{cfg}});
```

Comment:
The `protocol` and `host` properties are not used in your update function, so this can be simplified:

```suggestion
    local %ddclient::config = %{$tc->{cfg}};
```